File: frontend/src/config/orchestrationPresets.js
export const ORCHESTRATION_MODES = ['standard', 'extended', 'max', 'custom']

export const ORCHESTRATION_PRESETS = {
  standard: {
    maxToolIterations: 20,
    maxToolContinuationPhases: 3,
    longTaskToolThreshold: 8,
    promptBeforeLongTasks: true
  },
  extended: {
    maxToolIterations: 30,
    maxToolContinuationPhases: 4,
    longTaskToolThreshold: 10,
    promptBeforeLongTasks: true
  },
  max: {
    maxToolIterations: 40,
    maxToolContinuationPhases: 5,
    longTaskToolThreshold: 12,
    promptBeforeLongTasks: true
  }
}

export const ORCHESTRATION_MODE_OPTIONS = [
  {
    value: 'standard',
    label: 'Standard',
    description: 'Balanced for everyday operator tasks and short deployments.'
  },
  {
    value: 'extended',
    label: 'Extended',
    description: 'More tool rounds and continuations for multi-step configuration work.'
  },
  {
    value: 'max',
    label: 'Max',
    description: 'Highest limits for complex HA upgrades and long deployment chains.'
  },
  {
    value: 'custom',
    label: 'Custom',
    description: 'Set your own limits within the supported ranges.'
  }
]

export function inferOrchestrationMode(settings) {
  const stored = settings?.orchestrationMode
  if (stored && ORCHESTRATION_MODES.includes(stored)) {
    return stored
  }
  for (const mode of ['standard', 'extended', 'max']) {
    const preset = ORCHESTRATION_PRESETS[mode]
    const iterations = Number(settings?.maxToolIterations ?? preset.maxToolIterations)
    const phases = Number(settings?.maxToolContinuationPhases ?? preset.maxToolContinuationPhases)
    const threshold = Number(settings?.longTaskToolThreshold ?? preset.longTaskToolThreshold)
    const prompt = settings?.promptBeforeLongTasks ?? preset.promptBeforeLongTasks
    if (
      iterations === preset.maxToolIterations &&
      phases === preset.maxToolContinuationPhases &&
      threshold === preset.longTaskToolThreshold &&
      Boolean(prompt) === preset.promptBeforeLongTasks
    ) {
      return mode
    }
  }
  return 'custom'
}

export function orchestrationModeDescription(mode) {
  return ORCHESTRATION_MODE_OPTIONS.find((option) => option.value === mode)?.description || ''
}

export function applyOrchestrationPreset(mode) {
  if (!ORCHESTRATION_PRESETS[mode]) return null
  return { ...ORCHESTRATION_PRESETS[mode] }
}

/** Max model↔tool loops in one chat turn: (continuation phases + 1) × rounds per phase. */
export function effectiveMaxToolRounds(settings) {
  const roundsPerPhase = Number(settings?.maxToolIterations ?? 0)
  const continuationPhases = Number(settings?.maxToolContinuationPhases ?? 0)
  if (!roundsPerPhase || continuationPhases < 0) return 0
  return (continuationPhases + 1) * roundsPerPhase
}

export function effectiveMaxToolRoundsLabel(settings) {
  const total = effectiveMaxToolRounds(settings)
  const roundsPerPhase = Number(settings?.maxToolIterations ?? 0)
  const continuationPhases = Number(settings?.maxToolContinuationPhases ?? 0)
  const phaseCount = continuationPhases + 1
  return `${total} (${phaseCount} × ${roundsPerPhase})`
}

File: backend-api/app/services/orchestration_presets.py
"""Preset agent orchestration profiles for platform settings."""

from __future__ import annotations

from typing import Any, Literal

OrchestrationMode = Literal["standard", "extended", "max", "custom"]

ORCHESTRATION_PRESET_MODES: tuple[str, ...] = ("standard", "extended", "max")

ORCHESTRATION_PRESETS: dict[str, dict[str, Any]] = {
    "standard": {
        "maxToolIterations": 20,
        "maxToolContinuationPhases": 3,
        "longTaskToolThreshold": 8,
        "promptBeforeLongTasks": True,
    },
    "extended": {
        "maxToolIterations": 30,
        "maxToolContinuationPhases": 4,
        "longTaskToolThreshold": 10,
        "promptBeforeLongTasks": True,
    },
    "max": {
        "maxToolIterations": 40,
        "maxToolContinuationPhases": 5,
        "longTaskToolThreshold": 12,
        "promptBeforeLongTasks": True,
    },
}

DEFAULT_ORCHESTRATION_MODE: OrchestrationMode = "standard"


def preset_values(mode: str) -> dict[str, Any] | None:
    return ORCHESTRATION_PRESETS.get(mode)


def infer_orchestration_mode(document: dict[str, Any] | None) -> OrchestrationMode:
    document = document or {}
    stored = str(document.get("orchestrationMode") or "").strip().lower()
    if stored in ORCHESTRATION_PRESETS:
        return stored  # type: ignore[return-value]
    if stored == "custom":
        return "custom"

    for mode, preset in ORCHESTRATION_PRESETS.items():
        if all(document.get(key, value) == value for key, value in preset.items()):
            return mode  # type: ignore[return-value]
    return "custom"
